cmd/compute-domain-kubelet-plugin/driver.go

```diff
@@ -47,10 +49,93 @@ func isPermanentError(err error) bool {
 }
 
 type driver struct {
-	sync.Mutex
 	client       coreclientset.Interface
 	pluginhelper *kubeletplugin.Helper
 	state        *DeviceState
+	pulock       *Flock
+}
+
+type Flock struct {
+	path       string
+	PollPeriod time.Duration
+}
+
+func NewFlock(path string) *Flock {
+	return &Flock{
+		path: path,
+		// Short default period to keep lock acquisition rather responsive;
+		// adjust on a per-use case basis.
+		PollPeriod: 200 * time.Millisecond,
+	}
+}
+
+// Acquire attempts to acquire an exclusive file lock, polling with the configured
+// PollPeriod until whichever comes first:
+//
+// - lock successfully acquired
+// - timeout (if provided)
+// - external cancellation of context
+//
+// Returns a release function that must be called to unlock the file, typically
+// with defer().
+//
+// Introduced to protect the work in nodePrepareResource() and
+// nodeUnprepareResource() under a file-based lock because more than one driver
+// pod may be running on a node, but at most one such function must execute at
+// any given time.
+func (l *Flock) Acquire(ctx context.Context, timeout ...time.Duration) (func(), error) {
```

Comment:
We should use the functional options pattern for optional arguments instead of what you are doing here. With this, we could also make the polling period an optional parameter this way as well on a per-aquire() basis (and use the default otherwise).

https://dev.to/tiaguinho/functional-options-pattern-in-go-4e7m

Reply:
done